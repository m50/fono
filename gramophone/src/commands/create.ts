import { join } from 'path';
import upperFirst from 'lodash/upperFirst';
import camelCase from 'lodash/camelCase';
import { writeFile } from 'fs/promises';

export default async (tablename: string) => {
  const table = tablename.toLowerCase();
  const typename = upperFirst(camelCase(table));
  const timestamp = Date.now();
  const template = `import db from '../setup/db';

export const timestamp = ${timestamp};

export const up = async () => {
  await db.schema.createTable('${table}', (table) => {
    table.increments();

    table.timestamp('createdAt').defaultTo(db.fn.now());
    table.timestamp('updatedAt').defaultTo(db.fn.now());
  });
};

export const down = async () => {
  await db.schema.dropTableIfExists('${table}');
};

export interface ${typename} {
  id: number;

  createdAt: string;
  updatedAt: string;
}
`;
  const path = join(__dirname, '..', 'schema', `${typename}.ts`);
  await writeFile(path, template);
};
