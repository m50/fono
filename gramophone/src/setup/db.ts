import knex from 'knex';
import { DateTime } from 'luxon';

type Result = Record<string, any> | Record<string, any>[] | undefined | null;

function castTimestamps(result?: Record<string, any>) {
  if (!result) {
    return result;
  }
  const newResult = result;
  Object.keys(newResult).forEach((key) => {
    if (typeof newResult[key] !== 'string') {
      return;
    }
    const time = DateTime.fromISO(newResult[key]);
    if (time.isValid) {
      newResult[key] = time;
    }
  });
  return newResult;
}
const testConfig = {
  client: 'sqlite3',
  connection: {
    filename: ':memory:',
  },
  useNullAsDefault: true,
};

const prodConfig = {
  client: 'mysql',
  connection: {
    host: process.env.DB_HOST,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
  },
};

export default knex({
  ...(process.env.NODE_ENV === 'test' ? testConfig : prodConfig),
  postProcessResponse: (result: Result) => {
    if (!result) {
      return result;
    }

    if (Array.isArray(result)) {
      return result.map(castTimestamps);
    }

    return castTimestamps(result);
  },
});
