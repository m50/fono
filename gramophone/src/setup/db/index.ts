import { knex, Knex } from 'knex';
import castTimestamps from './castTimestamps';
import getRelationships from './relations';

type Result = Record<string, any> | Record<string, any>[] | undefined | null;

const config: Record<string, Knex.Config> = {
  test: {
    client: 'sqlite3',
    connection: {
      filename: ':memory:',
    },
    useNullAsDefault: true,
  },
  production: {
    client: 'mysql',
    connection: {
      host: process.env.DB_HOST,
      user: process.env.DB_USER,
      password: process.env.DB_PASSWORD,
      database: process.env.DB_DATABASE,
    },
  },
};

const db = knex({
  ...(config[process.env.NODE_ENV ?? 'test'] ?? config.production),
  postProcessResponse(result: Result, context: Record<string, any>) {
    if (!result || (typeof result !== 'object' && !Array.isArray(result))) {
      return result;
    }

    if (Array.isArray(result)) {
      return result.map((record) => getRelationships(castTimestamps(record), context));
    }

    return getRelationships(castTimestamps(result), context);
  },
});

knex.QueryBuilder.extend('withRelations', function withRelations(...relations: string[]) {
  // eslint-disable-next-line
  const { table } = this._single;
  return this.queryContext({ relations, db, table });
});

export default db;
