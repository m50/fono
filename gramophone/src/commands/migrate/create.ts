import { join } from 'path';
import { snakeCase, camelCase, upperFirst } from 'lodash';
import { writeFile } from 'fs/promises';
import { existsSync } from 'fs';
import { singular, plural } from 'pluralize';
import { exec } from 'child_process';
import chalk from 'chalk';

function open(path: string) {
  if (process.platform === 'darwin' && process.env.NODE_ENV !== 'test') {
    exec(`open "${path}"`);
  }
}

export default async (tablename: string) => {
  const table = plural(snakeCase(tablename)).toLowerCase();
  const typename = upperFirst(camelCase(singular(table)));
  const timestamp = Date.now();
  const template = `import { DateTime } from 'luxon';
import db from 'setup/db';

export const timestamp = ${timestamp};

export const ${plural(typename)} = () => db<${typename}>('${table}');

export interface ${typename} {
  id: number;

  createdAt: DateTime;
  updatedAt: DateTime;
}

export const up = async () => {
  await db.schema.createTable('${table}', (table) => {
    table.increments();

    table.timestamp('createdAt').defaultTo(db.fn.now());
    table.timestamp('updatedAt').defaultTo(db.fn.now());
  });
};

export const down = async () => {
  await db.schema.dropTableIfExists('${table}');
};
`;
  const path = join(__dirname, '..', '..', 'schema', `${typename}.ts`);
  if (existsSync(path)) {
    console.log(chalk.red('\nMigration already exists.\n'));
    open(path);
    return;
  }
  await writeFile(path, template);
  const fname = chalk.cyan(`${typename}.ts`);
  const tname = chalk.cyan(table);
  console.log(`\nCreated migration ${fname} for table \`${tname}\` \n\tat ${chalk.dim(path)}\n`);
  open(path);
};
