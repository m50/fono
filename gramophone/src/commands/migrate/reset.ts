import chalk from 'chalk';
import { readdir } from 'fs/promises';
import { join } from 'path';
import db from '../../setup/db';
import { Migration } from './types';

export default async () => {
  if (!(await db.schema.hasTable('migrations'))) {
    console.log('Nothing to reset.');
    return;
  }

  const dir = join(__dirname, '..', '..', 'schema');
  const files = await readdir(dir);
  // eslint-disable-next-line
  const promises = files.map((file) => require(`${dir}/${file}`) as Migration)
    .sort((a: Migration, b: Migration) => (a.timestamp > b.timestamp ? -1 : 1))
    .map(async (schema: Migration) => {
      const { down } = schema;
      await down();
    });
  await Promise.all(promises);
  await db.schema.dropTableIfExists('migrations');
  console.log(chalk.yellow('\n\t⚠️  Rolled back all migrations. ⚠️\n'));
};
