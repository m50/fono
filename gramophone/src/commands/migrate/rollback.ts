import { readdir } from 'fs/promises';
import { join } from 'path';
import { sha256 } from 'utils/hash';
import db from '../../setup/db';
import { Migration } from './types';

export default async () => {
  if (!(await db.schema.hasTable('migrations'))) {
    return;
  }

  const eventId = await db('migrations').max('eventId as eid').first()
    .then((v) => (v?.eid ?? -1) as number);

  if (eventId === -1) {
    return;
  }

  const dir = join(__dirname, '..', '..', 'schema');
  const files = await readdir(dir);
  // eslint-disable-next-line
  const promises = files.map((file) => require(`${dir}/${file}`))
    .sort((a: Migration, b: Migration) => (a.timestamp > b.timestamp ? 1 : -1))
    .map(async (schema) => {
      const { down, timestamp } = schema;
      const migrationId = sha256(timestamp);
      const exists = await db('migrations').where('migrationId', migrationId)
        .then((migrations) => migrations.length > 0);
      if (!exists) {
        return;
      }
      await down();
      await db('migrations')
        .where('migrationId', migrationId)
        .where('eventId', eventId)
        .delete();
    });
  await Promise.all(promises);
};
