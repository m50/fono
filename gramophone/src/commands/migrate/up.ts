import { readdir } from 'fs/promises';
import { join } from 'path';
import { sha256 } from 'utils/hash';
import db from '../../setup/db';
import { MigrationFile } from './types';

export default async () => {
  if (!(await db.schema.hasTable('migrations'))) {
    await db.schema.createTable('migrations', (table) => {
      table.string('migrationId').primary().unique();
      table.integer('eventId');
      table.timestamp('timestamp').defaultTo(db.fn.now());
    });
  }

  const eventId = await db('migrations').max('eventId as eid').first()
    .then((v) => (v?.eid ?? -1) as number)
    .then((eid) => eid + 1);

  const dir = join(__dirname, '..', '..', 'schema');
  const files = await readdir(dir);
  // eslint-disable-next-line
  const promises = files.map((file) => [file, require(`${dir}/${file}`)] as MigrationFile)
    .sort((a: MigrationFile, b: MigrationFile) => (a[1].timestamp > b[1].timestamp ? 1 : -1))
    .map(async (migrationFile: MigrationFile) => {
      const [file, schema] = migrationFile;
      const { up, timestamp } = schema;
      const migrationId = sha256(timestamp);
      const exists = await db('migrations').where('migrationId', migrationId)
        .then((migrations) => migrations.length > 0);
      if (exists) {
        return;
      }
      await up();
      await db('migrations').insert({ migrationId, eventId });
      console.log(`Migrated ${file} [${timestamp} round ${eventId}]`);
    });
  await Promise.all(promises);
};
