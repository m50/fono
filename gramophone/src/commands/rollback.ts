import { readdir } from 'fs/promises';
import { join } from 'path';
import db from '../setup/db';

export default async () => {
  if (!(await db.schema.hasTable('migrations'))) {
    console.log('Nothing to rollback!');
    return;
  }

  const eventId = await db('migrations').max('eventId as eid').first()
    .then((v) => (v?.eid ?? -1) as number);

  if (eventId === -1) {
    console.log('Nothing to rollback!');
    return;
  }

  const dir = join(__dirname, '..', 'schema');
  const files = await readdir(dir);
  // eslint-disable-next-line
  const promises = files.map((file) => require(`${dir}/${file}`))
    .map(async (schema) => {
      const { down, timestamp } = schema;
      const exists = await db('migrations').where('migrationId', timestamp)
        .then((migrations) => migrations.length > 0);
      if (!exists) {
        return;
      }
      await down();
      await db('migrations')
        .where('migrationId', timestamp)
        .where('eventId', eventId)
        .delete();
    });
  await Promise.all(promises);
};
