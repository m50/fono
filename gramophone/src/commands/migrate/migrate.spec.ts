import { readFile, rm } from 'fs/promises';
import { existsSync } from 'fs';
import { join } from 'path';
import up from './up';
import rollback from './rollback';
import reset from './reset';
import create from './create';
import db from '../../setup/db';

const { log } = console;

describe('migrate', () => {
  beforeAll(() => { console.log = jest.fn(); });
  afterAll(() => { console.log = log; });

  afterEach(() => rollback());

  describe('up', () => {
    it('can run migrations', async () => {
      await up();
      const count = await db('migrations').count('migrationId');
      expect(count[0]['count(`migrationId`)']).toBeGreaterThan(0);
      expect(console.log).toHaveBeenCalled();
    });
  });
  describe('rollback', () => {
    it('can rollback migrations', async () => {
      await up();
      const count = await db('migrations').count('migrationId');
      expect(count[0]['count(`migrationId`)']).toBeGreaterThan(0);
      expect(console.log).toHaveBeenCalled();

      await rollback();
      const count2 = await db('migrations').count('migrationId');
      expect(count2[0]['count(`migrationId`)']).toBe(0);
      expect(console.log).toHaveBeenCalled();
    });
  });
  describe('reset', () => {
    it('empties database', async () => {
      await up();
      const count = await db('migrations').count('migrationId');
      expect(count[0]['count(`migrationId`)']).toBeGreaterThan(0);
      await reset();
      const migrationsExist = await db.schema.hasTable('migrations');
      expect(migrationsExist).toBe(false);
    });
  });

  describe('create', () => {
    it('creates a migration file with snake_case param', async () => {
      Date.now = jest.fn(() => 1618196206380);
      await create('migration_test');
      const path = join(__dirname, '..', '..', 'schema', 'MigrationTest.ts');
      expect(existsSync(path)).toBeTruthy();
      const fileData = await readFile(path);
      const file = fileData.toString();
      expect(file).toMatch(/export interface MigrationTest/);
      expect(file).toMatch(/export const MigrationTests = \(\) => db<MigrationTest>\('migration_tests'\);/);
      expect(file).toMatch(/createdAt: DateTime;/);
      expect(file).toMatch(/updatedAt: DateTime;/);
      expect(file).toMatch(/await db\.schema\.createTable\('migration_tests'/);
      expect(file).toMatch(/export const up = async/);
      expect(file).toMatch(/export const down = async/);
      expect(file).toMatch(/export const timestamp = \d+;/);
      expect(file).toMatchSnapshot();
      expect(console.log).toHaveBeenCalled();
      await rm(path);
    });

    it('creates a migration file with PascalCase param', async () => {
      Date.now = jest.fn(() => 1618196206380);
      await create('MigrationTest');
      const path = join(__dirname, '..', '..', 'schema', 'MigrationTest.ts');
      expect(existsSync(path)).toBeTruthy();
      const fileData = await readFile(path);
      const file = fileData.toString();
      expect(file).toMatch(/export interface MigrationTest/);
      expect(file).toMatch(/export const MigrationTests = \(\) => db<MigrationTest>\('migration_tests'\);/);
      expect(file).toMatch(/createdAt: DateTime;/);
      expect(file).toMatch(/updatedAt: DateTime;/);
      expect(file).toMatch(/await db\.schema\.createTable\('migration_tests'/);
      expect(file).toMatch(/export const up = async/);
      expect(file).toMatch(/export const down = async/);
      expect(file).toMatch(/export const timestamp = \d+;/);
      expect(file).toMatchSnapshot();
      await rm(path);
    });
  });
});
