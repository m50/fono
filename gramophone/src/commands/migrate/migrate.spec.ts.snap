// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrate create creates a migration file with PascalCase param 1`] = `
"import { DateTime } from 'luxon';
import db from 'setup/db';

export const timestamp = 1618196206380;

export const MigrationTests = () => db<MigrationTest>('migration_tests');

export interface MigrationTest {
  id: number;

  createdAt: DateTime;
  updatedAt: DateTime;
}

export const up = async () => {
  await db.schema.createTable('migration_tests', (table) => {
    table.increments();

    table.timestamp('createdAt').defaultTo(db.fn.now());
    table.timestamp('updatedAt').defaultTo(db.fn.now());
  });
};

export const down = async () => {
  await db.schema.dropTableIfExists('migration_tests');
};
"
`;

exports[`migrate create creates a migration file with snake_case param 1`] = `
"import { DateTime } from 'luxon';
import db from 'setup/db';

export const timestamp = 1618196206380;

export const MigrationTests = () => db<MigrationTest>('migration_tests');

export interface MigrationTest {
  id: number;

  createdAt: DateTime;
  updatedAt: DateTime;
}

export const up = async () => {
  await db.schema.createTable('migration_tests', (table) => {
    table.increments();

    table.timestamp('createdAt').defaultTo(db.fn.now());
    table.timestamp('updatedAt').defaultTo(db.fn.now());
  });
};

export const down = async () => {
  await db.schema.dropTableIfExists('migration_tests');
};
"
`;
